File: step7/scripts/plot_homozygosity.py
#!/usr/bin/env python3
##########################################################################
# CRISP - Comprehensive Robust Integrated SNP Processing
# Step 6: Heterozygosity and Homozygosity, Plotting Script
# Version: 0.2.0
# Developed by Igor Pupko
# https://github.com/ipupko/CRISP
# Part of the Compass Genomics suite
##########################################################################

import sys
import os
import argparse
import pandas as pd
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.ticker as ticker


# ─────────────────────────────────────────────
# CONSTANTS
# ─────────────────────────────────────────────
COLORS = {
    'Normal'              : '#1D9E75',
    'High Heterozygosity' : '#febc2e',
    'Excess Homozygosity' : '#ff5f57',
}

CAPTION = "CRISP | Comprehensive Robust Integrated SNP Processing"


# ─────────────────────────────────────────────
# LOGGING
# ─────────────────────────────────────────────
def log(msg: str):
    print(f"[HOMO] {msg}", flush=True)

def abort(msg: str):
    print(f"[HOMO] ERROR: {msg}", file=sys.stderr, flush=True)
    sys.exit(1)


# ─────────────────────────────────────────────
# LOAD FILES
# ─────────────────────────────────────────────
def load_file(filepath: str, label: str) -> pd.DataFrame:
    if not os.path.isfile(filepath):
        abort(f"{label} not found: {filepath}")
    df = pd.read_csv(filepath, sep=r'\s+')
    log(f"Loaded {label}: {len(df):,} rows")
    return df


# ─────────────────────────────────────────────
# PREPARE DATA
# ─────────────────────────────────────────────
def prepare_data(df_sex: pd.DataFrame,
                 df_het: pd.DataFrame,
                 df_hom: pd.DataFrame) -> pd.DataFrame:
    """
    Merge sexcheck, het and hom.indiv into a single DataFrame.
    Calculates Z-scores for autosomal and X chromosome F-statistics.
    """
    # Sex index from sexcheck
    sex_index = df_sex[['IID', 'PEDSEX']].copy()

    # Merge sex into het
    df_het = df_het.merge(sex_index, on='IID', how='left')

    # Calculate autosomal Z-score
    f_mean = df_het['F'].mean()
    f_sd   = df_het['F'].std()
    df_het['Z_Score'] = (df_het['F'] - f_mean) / f_sd

    log(f"Autosomal F-stat mean : {f_mean:.4f}")
    log(f"Autosomal F-stat SD   : {f_sd:.4f}")

    # Merge with ROH data
    df_merged = df_het.merge(
        df_hom[['IID', 'KB', 'NSEG']],
        on  = 'IID',
        how = 'left'
    )

    return df_merged, f_mean, f_sd


# ─────────────────────────────────────────────
# CLASSIFY OUTLIERS
# ─────────────────────────────────────────────
def classify(df: pd.DataFrame,
             z_high: float,
             z_low: float) -> pd.DataFrame:
    df = df.copy()
    df['Category'] = 'Normal'
    df.loc[df['Z_Score'] > z_high, 'Category'] = 'Excess Homozygosity'
    df.loc[df['Z_Score'] < z_low,  'Category'] = 'High Heterozygosity'
    return df


# ─────────────────────────────────────────────
# PLOT
# ─────────────────────────────────────────────
def generate_plot(df: pd.DataFrame,
                  z_high: float,
                  z_low: float,
                  output_dir: str) -> str:

    n_excess = (df['Category'] == 'Excess Homozygosity').sum()
    n_het    = (df['Category'] == 'High Heterozygosity').sum()
    n_normal = (df['Category'] == 'Normal').sum()

    fig, ax = plt.subplots(figsize=(11, 7))

    for cat, grp in df.groupby('Category'):
        ax.scatter(
            grp['KB'], grp['Z_Score'],
            c      = COLORS[cat],
            s      = 18,
            alpha  = 0.75,
            label  = cat,
            zorder = 2
        )

    # Threshold lines
    ax.axhline(0,      color='black',   linestyle='dashed',
               linewidth=0.6, alpha=0.5)
    ax.axhline(z_high, color='#ff5f57', linestyle='dashed',
               linewidth=1.0, alpha=0.8)
    ax.axhline(z_low,  color='#febc2e', linestyle='dashed',
               linewidth=1.0, alpha=0.8)

    # Threshold annotations
    xlim = ax.get_xlim()
    xpos = xlim[0] + (xlim[1] - xlim[0]) * 0.02

    ax.text(xpos, z_high + 0.12,
            f'Z = {z_high} (excess homozygosity)',
            color='#ff5f57', fontsize=8, fontweight='bold')
    ax.text(xpos, z_low - 0.18,
            f'Z = {z_low} (high heterozygosity)',
            color='#febc2e', fontsize=8, fontweight='bold')

    # Axes formatting
    ax.xaxis.set_major_formatter(
        ticker.FuncFormatter(lambda x, _: f'{x:,.0f}'))
    ax.set_xlabel('Total runs of homozygosity (KB)',
                  fontweight='bold', fontsize=10)
    ax.set_ylabel('Homozygosity Z-score',
                  fontweight='bold', fontsize=10)
    ax.set_title(
        'CRISP: Step 6: Total Homozygosity Runs vs Homozygosity Z-score',
        fontweight='bold', fontsize=12
    )
    fig.suptitle(
        f'Z > {z_high}: {n_excess} excess homozygosity  |  '
        f'Z < {z_low}: {n_het} high heterozygosity  |  '
        f'{n_normal} normal',
        fontsize=9, color='#555555', y=0.96
    )

    # Style
    ax.set_facecolor('#f8f9fa')
    fig.patch.set_facecolor('white')
    ax.spines[['top', 'right']].set_visible(False)
    ax.tick_params(labelsize=9)

    # Legend
    patches = [mpatches.Patch(color=COLORS[c], label=c)
               for c in COLORS]
    ax.legend(handles=patches, loc='upper right',
              fontsize=9, framealpha=0.9)

    fig.text(0.5, 0.01, CAPTION,
             ha='center', fontsize=8, color='#888888')
    plt.tight_layout(rect=[0, 0.03, 1, 0.95])

    out_file = os.path.join(
        output_dir, "Homozygosity_Runs_vs_Zscore.pdf"
    )
    fig.savefig(out_file, bbox_inches='tight', dpi=150)
    plt.close(fig)

    log(f"Plot saved: {out_file}")
    return out_file


# ─────────────────────────────────────────────
# WRITE REPORT
# ─────────────────────────────────────────────
def write_report(df: pd.DataFrame,
                 z_high: float,
                 z_low: float,
                 f_mean: float,
                 f_sd: float,
                 output_dir: str) -> str:

    outliers_excess = df[df['Category'] == 'Excess Homozygosity']
    outliers_het    = df[df['Category'] == 'High Heterozygosity']

    report_path = os.path.join(output_dir, "report_homozygosity.txt")

    with open(report_path, 'w') as out:
        out.write("=" * 64 + "\n")
        out.write("  CRISP: STEP 6 HETEROZYGOSITY AND HOMOZYGOSITY REPORT\n")
        out.write("  Comprehensive Robust Integrated SNP Processing\n")
        out.write("=" * 64 + "\n\n")
        out.write(f"  Z-score upper : {z_high} (excess homozygosity)\n")
        out.write(f"  Z-score lower : {z_low} (high heterozygosity)\n\n")

        out.write("-" * 64 + "\n")
        out.write("AUTOSOMAL HETEROZYGOSITY\n")
        out.write("-" * 64 + "\n")
        out.write(f"  Samples total          : {len(df):,}\n")
        out.write(f"  F-stat mean            : {f_mean:.4f}\n")
        out.write(f"  F-stat SD              : {f_sd:.4f}\n")
        out.write(f"  Excess homozygosity    : {len(outliers_excess):,} "
                  f"(Z > {z_high})\n")
        out.write(f"  High heterozygosity    : {len(outliers_het):,} "
                  f"(Z < {z_low})\n\n")

        out.write("-" * 64 + "\n")
        out.write("EXCESS HOMOZYGOSITY SAMPLES\n")
        out.write("-" * 64 + "\n")
        if len(outliers_excess) > 0:
            out.write(outliers_excess[
                ['FID', 'IID', 'F', 'Z_Score', 'KB']
            ].to_string(index=False))
            out.write("\n\n")
        else:
            out.write("  None detected.\n\n")

        out.write("-" * 64 + "\n")
        out.write("HIGH HETEROZYGOSITY SAMPLES\n")
        out.write("-" * 64 + "\n")
        if len(outliers_het) > 0:
            out.write(outliers_het[
                ['FID', 'IID', 'F', 'Z_Score', 'KB']
            ].to_string(index=False))
            out.write("\n\n")
        else:
            out.write("  None detected.\n\n")

        out.write("=" * 64 + "\n")
        out.write("  END OF REPORT\n")
        out.write("=" * 64 + "\n")

    log(f"Report written: {report_path}")
    return report_path


# ─────────────────────────────────────────────
# WRITE EXCLUSION LISTS
# ─────────────────────────────────────────────
def write_exclusion_lists(df: pd.DataFrame, output_dir: str):

    outliers_excess = df[df['Category'] == 'Excess Homozygosity']
    outliers_het    = df[df['Category'] == 'High Heterozygosity']

    # High heterozygosity
    het_file = os.path.join(output_dir, "outliers_high_het.txt")
    outliers_het[['FID', 'IID']].to_csv(
        het_file, sep='\t', index=False, header=False
    )
    log(f"High het exclusion list     : {het_file} "
        f"({len(outliers_het):,} samples)")

    # Excess homozygosity
    homo_file = os.path.join(output_dir, "outliers_excess_homo.txt")
    outliers_excess[['FID', 'IID']].to_csv(
        homo_file, sep='\t', index=False, header=False
    )
    log(f"Excess homo exclusion list  : {homo_file} "
        f"({len(outliers_excess):,} samples)")

    # Combined
    combined = pd.concat(
        [outliers_het[['FID', 'IID']],
         outliers_excess[['FID', 'IID']]]
    ).drop_duplicates()

    combined_file = os.path.join(output_dir, "outliers_combined.txt")
    combined.to_csv(
        combined_file, sep='\t', index=False, header=False
    )
    log(f"Combined exclusion list     : {combined_file} "
        f"({len(combined):,} samples)")

    return het_file, homo_file, combined_file


# ─────────────────────────────────────────────
# MAIN
# ─────────────────────────────────────────────
def main():
    parser = argparse.ArgumentParser(
        description="CRISP Step 6: Heterozygosity and Homozygosity (Python)"
    )
    parser.add_argument("output_dir")
    parser.add_argument("sexcheck_file")
    parser.add_argument("het_file")
    parser.add_argument("hom_indiv_file")
    parser.add_argument("z_high", type=float)
    parser.add_argument("z_low",  type=float)
    args = parser.parse_args()

    os.makedirs(args.output_dir, exist_ok=True)

    log("CRISP Step 6: Heterozygosity and Homozygosity (Python)")
    log(f"Z-score upper threshold : {args.z_high} (excess homozygosity)")
    log(f"Z-score lower threshold : {args.z_low} (high heterozygosity)")
    log("")

    # Load files
    df_sex = load_file(args.sexcheck_file, ".sexcheck")
    df_het = load_file(args.het_file,      ".het")
    df_hom = load_file(args.hom_indiv_file, ".hom.indiv")
    log("")

    # Prepare and merge
    log("Preparing data...")
    df_merged, f_mean, f_sd = prepare_data(df_sex, df_het, df_hom)
    log("")

    # Classify outliers
    log("Classifying outliers...")
    df_classified = classify(df_merged, args.z_high, args.z_low)

    n_excess = (df_classified['Category'] == 'Excess Homozygosity').sum()
    n_het    = (df_classified['Category'] == 'High Heterozygosity').sum()
    n_normal = (df_classified['Category'] == 'Normal').sum()

    log(f"  Excess homozygosity (Z > {args.z_high}) : {n_excess:,}")
    log(f"  High heterozygosity (Z < {args.z_low}) : {n_het:,}")
    log(f"  Normal                                  : {n_normal:,}")
    log("")

    # Generate plot
    log("Generating plot...")
    generate_plot(df_classified, args.z_high, args.z_low,
                  args.output_dir)
    log("")

    # Write report
    log("Writing report...")
    write_report(df_classified, args.z_high, args.z_low,
                 f_mean, f_sd, args.output_dir)
    log("")

    # Write exclusion lists
    log("Writing exclusion lists...")
    write_exclusion_lists(df_classified, args.output_dir)

    log("\nHomozygosity check complete.")


if __name__ == "__main__":
    main()

File: step7/scripts/plot_homozygosity.R
#!/usr/bin/env Rscript
##########################################################################
### CRISP - Comprehensive Robust Integrated SNP Processing
### Step 6: Heterozygosity and Homozygosity, Plotting Script
### Version: 0.2.0
### Developed by Igor Pupko
### https://github.com/ipupko/CRISP
### Part of the Compass Genomics suite
##########################################################################
# Performs heterozygosity and homozygosity QC using PLINK output.
# Detects outliers using Z-score thresholds on autosomal F-statistics
# and visualises total runs of homozygosity (ROH) against Z-scores.
#
# Inputs:
#   .sexcheck file   : X chromosome F-statistics (sex check step)
#   .het file        : autosomal heterozygosity
#   .hom.indiv file  : runs of homozygosity per individual
#
# Outputs:
#   Homozygosity_Runs_vs_Zscore.pdf    ROH vs Z-score scatter plot
#   report_homozygosity.txt            Detailed text report
#   outliers_high_het.txt              High heterozygosity exclusions
#   outliers_excess_homo.txt           Excess homozygosity exclusions
#   outliers_combined.txt              Combined exclusion list
#
# Usage:
#   Rscript plot_homozygosity.R <output_dir> <sexcheck_file>
#                               <het_file> <hom_indiv_file>
#                               <z_high> <z_low>
##########################################################################

suppressPackageStartupMessages({
    library(ggplot2)
    library(data.table)
    library(scales)
})

args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 6) {
    cat("Usage: Rscript plot_homozygosity.R",
        "<output_dir> <sexcheck_file> <het_file>",
        "<hom_indiv_file> <z_high> <z_low>\n")
    quit(status = 1)
}

output_dir      <- args[1]
sexcheck_file   <- args[2]
het_file        <- args[3]
hom_indiv_file  <- args[4]
z_high          <- as.numeric(args[5])
z_low           <- as.numeric(args[6])

dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

cat(sprintf("[HOMO] Parameters:\n"))
cat(sprintf("  Z-score upper threshold : %.1f (excess homozygosity)\n", z_high))
cat(sprintf("  Z-score lower threshold : %.1f (high heterozygosity)\n", z_low))
cat("\n")

##########################################################################
### LOAD DATA
##########################################################################

cat(sprintf("[HOMO] Loading sexcheck file : %s\n", sexcheck_file))
x_homozygosity <- fread(sexcheck_file, header = TRUE)

cat(sprintf("[HOMO] Loading het file      : %s\n", het_file))
autosome_het <- fread(het_file, header = TRUE)

cat(sprintf("[HOMO] Loading hom.indiv    : %s\n", hom_indiv_file))
hom_indiv <- fread(hom_indiv_file, header = TRUE)

cat(sprintf("[HOMO] Samples loaded: %d\n\n", nrow(autosome_het)))

##########################################################################
### PREPARE DATA
##########################################################################

# Create sex index from sexcheck for merging
sex_index <- x_homozygosity[, c("IID", "PEDSEX"), with = FALSE]

# Merge sex into autosomal het
autosome_het <- merge(autosome_het, sex_index,
                      by = "IID", all = TRUE)

# Add suffixes to avoid column name clashes
colnames(autosome_het) <- paste0(colnames(autosome_het), "_auto")
colnames(x_homozygosity) <- paste0(colnames(x_homozygosity), "_XChr")

##########################################################################
### CALCULATE Z-SCORES
##########################################################################

# X chromosome Z-score
x_homozygosity$Z_Score_XChr <- (
    x_homozygosity$F_XChr - mean(x_homozygosity$F_XChr, na.rm = TRUE)
) / sd(x_homozygosity$F_XChr, na.rm = TRUE)

# Autosomal Z-score
autosome_het$Z_Score_auto <- (
    autosome_het$F_auto - mean(autosome_het$F_auto, na.rm = TRUE)
) / sd(autosome_het$F_auto, na.rm = TRUE)

cat(sprintf("[HOMO] Autosomal F-stat mean : %.4f\n",
            mean(autosome_het$F_auto, na.rm = TRUE)))
cat(sprintf("[HOMO] Autosomal F-stat SD   : %.4f\n",
            sd(autosome_het$F_auto, na.rm = TRUE)))
cat("\n")

##########################################################################
### MERGE WITH ROH DATA
##########################################################################

autosome_het_roh <- merge(autosome_het, hom_indiv,
                           by.x = "IID_auto",
                           by.y = "IID",
                           all  = TRUE)

##########################################################################
### CLASSIFY OUTLIERS
##########################################################################

autosome_het_roh$Category <- ifelse(
    autosome_het_roh$Z_Score_auto > z_high, "Excess Homozygosity",
    ifelse(
        autosome_het_roh$Z_Score_auto < z_low, "High Heterozygosity",
        "Normal"
    )
)

n_excess_homo <- sum(autosome_het_roh$Category == "Excess Homozygosity",
                     na.rm = TRUE)
n_high_het    <- sum(autosome_het_roh$Category == "High Heterozygosity",
                     na.rm = TRUE)
n_normal      <- sum(autosome_het_roh$Category == "Normal", na.rm = TRUE)

cat(sprintf("[HOMO] Excess homozygosity (Z > %.1f)  : %d\n",
            z_high, n_excess_homo))
cat(sprintf("[HOMO] High heterozygosity (Z < %.1f)  : %d\n",
            z_low, n_high_het))
cat(sprintf("[HOMO] Normal                           : %d\n\n",
            n_normal))

##########################################################################
### PLOT: ROH KB VS HOMOZYGOSITY Z-SCORE
##########################################################################

cat("[HOMO] Generating plots...\n")

pdf_file <- file.path(output_dir, "Homozygosity_Runs_vs_Zscore.pdf")
pdf(pdf_file, width = 10, height = 7)

category_colors <- c(
    "Excess Homozygosity" = "#ff5f57",
    "High Heterozygosity" = "#febc2e",
    "Normal"              = "#1D9E75"
)

p <- ggplot(autosome_het_roh,
            aes(x = KB, y = Z_Score_auto, colour = Category)) +
    geom_point(alpha = 0.7, size = 1.5) +
    geom_hline(yintercept = 0,      linetype = "dashed",
               colour = "black",    linewidth = 0.5) +
    geom_hline(yintercept = z_high, linetype = "dashed",
               colour = "#ff5f57",  linewidth = 0.8) +
    geom_hline(yintercept = z_low,  linetype = "dashed",
               colour = "#febc2e",  linewidth = 0.8) +
    annotate("text", x = max(autosome_het_roh$KB, na.rm = TRUE) * 0.05,
             y = z_high + 0.15,
             label = paste0("Z = ", z_high, " (excess homozygosity)"),
             colour = "#ff5f57", size = 3, hjust = 0) +
    annotate("text", x = max(autosome_het_roh$KB, na.rm = TRUE) * 0.05,
             y = z_low - 0.15,
             label = paste0("Z = ", z_low, " (high heterozygosity)"),
             colour = "#febc2e", size = 3, hjust = 0) +
    scale_colour_manual(values = category_colors,
                        name   = "Category") +
    scale_x_continuous(labels = label_comma()) +
    labs(
        title    = "CRISP: Step 6: Total Homozygosity Runs vs Homozygosity Z-score",
        subtitle = sprintf(
            "Z > %.1f: %d excess homozygosity  |  Z < %.1f: %d high heterozygosity  |  %d normal",
            z_high, n_excess_homo, z_low, n_high_het, n_normal
        ),
        x        = "Total runs of homozygosity (KB)",
        y        = "Homozygosity Z-score",
        caption  = "CRISP | Comprehensive Robust Integrated SNP Processing"
    ) +
    theme_minimal(base_size = 12) +
    theme(
        plot.title       = element_text(face = "bold", size = 13),
        plot.subtitle    = element_text(colour = "#555555", size = 9),
        axis.title       = element_text(face = "bold"),
        panel.grid.minor = element_blank(),
        legend.position  = "right",
        plot.caption     = element_text(colour = "#888888", size = 8)
    )

print(p)
dev.off()

cat(sprintf("[HOMO] Plot saved: %s\n\n", pdf_file))

##########################################################################
### WRITE REPORT
##########################################################################

report_file <- file.path(output_dir, "report_homozygosity.txt")
sink(report_file)

cat("================================================================\n")
cat("  CRISP: STEP 6 HETEROZYGOSITY AND HOMOZYGOSITY REPORT\n")
cat("  Comprehensive Robust Integrated SNP Processing\n")
cat("================================================================\n\n")

cat(sprintf("  Date              : %s\n", Sys.time()))
cat(sprintf("  Z-score upper     : %.1f (excess homozygosity)\n", z_high))
cat(sprintf("  Z-score lower     : %.1f (high heterozygosity)\n\n", z_low))

cat("----------------------------------------------------------------\n")
cat("AUTOSOMAL HETEROZYGOSITY\n")
cat("----------------------------------------------------------------\n")
cat(sprintf("  Samples total          : %d\n", nrow(autosome_het_roh)))
cat(sprintf("  F-stat mean            : %.4f\n",
            mean(autosome_het_roh$F_auto, na.rm = TRUE)))
cat(sprintf("  F-stat SD              : %.4f\n",
            sd(autosome_het_roh$F_auto, na.rm = TRUE)))
cat(sprintf("  Excess homozygosity    : %d (Z > %.1f)\n",
            n_excess_homo, z_high))
cat(sprintf("  High heterozygosity    : %d (Z < %.1f)\n\n",
            n_high_het, z_low))

cat("----------------------------------------------------------------\n")
cat("EXCESS HOMOZYGOSITY SAMPLES\n")
cat("----------------------------------------------------------------\n")
outliers_excess <- autosome_het_roh[
    autosome_het_roh$Category == "Excess Homozygosity", ]
if (nrow(outliers_excess) > 0) {
    print(as.data.frame(outliers_excess[,
        c("IID_auto", "F_auto", "Z_Score_auto", "KB")]))
} else {
    cat("  None detected.\n")
}

cat("\n----------------------------------------------------------------\n")
cat("HIGH HETEROZYGOSITY SAMPLES\n")
cat("----------------------------------------------------------------\n")
outliers_het <- autosome_het_roh[
    autosome_het_roh$Category == "High Heterozygosity", ]
if (nrow(outliers_het) > 0) {
    print(as.data.frame(outliers_het[,
        c("IID_auto", "F_auto", "Z_Score_auto", "KB")]))
} else {
    cat("  None detected.\n")
}

cat("\n================================================================\n")
cat("  END OF REPORT\n")
cat("================================================================\n")
sink()

cat(sprintf("[HOMO] Report written: %s\n\n", report_file))

##########################################################################
### WRITE EXCLUSION LISTS
##########################################################################

# High heterozygosity (your original naming convention preserved)
het_excl_file <- file.path(output_dir, "outliers_high_het.txt")
if (nrow(outliers_het) > 0) {
    write.table(outliers_het[, c("FID_auto", "IID_auto")],
                het_excl_file,
                quote     = FALSE,
                row.names = FALSE,
                sep       = "\t",
                col.names = FALSE)
} else {
    file.create(het_excl_file)
}
cat(sprintf("[HOMO] High het exclusion list     : %s (%d samples)\n",
            het_excl_file, nrow(outliers_het)))

# Excess homozygosity
homo_excl_file <- file.path(output_dir, "outliers_excess_homo.txt")
if (nrow(outliers_excess) > 0) {
    write.table(outliers_excess[, c("FID_auto", "IID_auto")],
                homo_excl_file,
                quote     = FALSE,
                row.names = FALSE,
                sep       = "\t",
                col.names = FALSE)
} else {
    file.create(homo_excl_file)
}
cat(sprintf("[HOMO] Excess homo exclusion list  : %s (%d samples)\n",
            homo_excl_file, nrow(outliers_excess)))

# Combined exclusion list
combined_file <- file.path(output_dir, "outliers_combined.txt")
all_outliers  <- unique(rbind(
    outliers_het[,    c("FID_auto", "IID_auto")],
    outliers_excess[, c("FID_auto", "IID_auto")]
))
write.table(all_outliers,
            combined_file,
            quote     = FALSE,
            row.names = FALSE,
            sep       = "\t",
            col.names = FALSE)
cat(sprintf("[HOMO] Combined exclusion list     : %s (%d samples)\n\n",
            combined_file, nrow(all_outliers)))

cat("[HOMO] Homozygosity check complete.\n")
